core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -619,30 +582,68 @@ public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit
         {
             Driver driver = driverFactory.createDriver(driverContext);
 
-            // record driver so other threads add unpartitioned sources can see the driver
-            // NOTE: this MUST be done before reading unpartitionedSources, so we see a consistent view of the unpartitioned sources
-            drivers.add(new WeakReference<>(driver));
-
             if (partitionedSplit != null) {
                 // TableScanOperator requires partitioned split to be added before the first call to process
                 driver.updateSplitAssignment(new SplitAssignment(partitionedSplit.getPlanNodeId(), ImmutableSet.of(partitionedSplit), true));
             }
 
-            // add unpartitioned sources
-            Optional<PlanNodeId> sourceId = driver.getSourceId();
-            if (sourceId.isPresent()) {
-                SplitAssignment splitAssignment = unpartitionedSplitAssignments.get(sourceId.get());
-                if (splitAssignment != null) {
-                    driver.updateSplitAssignment(splitAssignment);
-                }
-            }
-
             status.decrementPendingCreation(pipelineContext.getPipelineId());
             closeDriverFactoryIfFullyCreated();
 
+            if (driverFactory.getSourceId().isPresent() && partitionedSplit == null) {
+                driverReferences.add(new WeakReference<>(driver));
+                scheduleSplits();
+            }
+
             return driver;
         }
 
+        public void enqueueSplits(Set<ScheduledSplit> splits, boolean noMoreSplits)
+        {
+            verify(driverFactory.getSourceId().isPresent(), "not a source driver");
+            queuedSplits.addAll(splits);
+            this.noMoreSplits.set(noMoreSplits);
+        }
+
+        public void scheduleSplits()
+        {
+            if (driverReferences.isEmpty()) {
+                return;
+            }
+
+            PlanNodeId sourceId = driverFactory.getSourceId().orElseThrow();
+            while (!queuedSplits.isEmpty()) {
+                int activeDriversCount = 0;
+                for (WeakReference<Driver> driverReference : driverReferences) {
+                    Driver driver = driverReference.get();
+                    if (driver == null) {
+                        continue;
```

Comment:
remove reference from the queue?

Reply:
This is a `CopyOnWriteArrayList`. The iterator doesn't support `remove`. Removing it from the original list is possible, but it will make it harder to reason about (what is already quite difficult to be honest). The `WeakReference` objects are relatively lightweight, not sure if the saved memory is worth the extra complexity.